graph/src/data/schema.rs

```diff
@@ -73,8 +73,204 @@ pub enum SchemaValidationError {
+            invalid => Err(format!(
+                "Provided language for fulltext search is invalid: {}",
+                invalid
+            )),
+        }
+    }
+}
+
+impl FulltextLanguage {
+    pub fn as_sql(&self) -> &'static str {
+        match self {
+            Self::Simple => "simple",
+            Self::Danish => "danish",
+            Self::Dutch => "dutch",
+            Self::English => "english",
+            Self::Finnish => "finnish",
+            Self::French => "french",
+            Self::German => "german",
+            Self::Hungarian => "hungarian",
+            Self::Italian => "italian",
+            Self::Norwegian => "norwegian",
+            Self::Portugese => "portugese",
+            Self::Romanian => "romanian",
+            Self::Russian => "russian",
+            Self::Spanish => "spanish",
+            Self::Swedish => "swedish",
+            Self::Turkish => "turkish",
+        }
+    }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub enum FulltextAlgorithm {
+    Ranked,
+    ProximityRanked,
+}
+
+impl TryFrom<&String> for FulltextAlgorithm {
+    type Error = String;
+    fn try_from(algorithm: &String) -> Result<Self, Self::Error> {
+        match &algorithm[..] {
+            "ranked" => Ok(FulltextAlgorithm::Ranked),
+            "proximity_ranked" => Ok(FulltextAlgorithm::ProximityRanked),
+            invalid => Err(format!(
+                "Provided algorithm for fulltext search is invalid: {}",
+                invalid,
+            )),
+        }
+    }
+}
+
+impl FulltextAlgorithm {
+    pub fn as_sql(&self) -> &'static str {
+        match self {
+            Self::Ranked => "ts_rank(",
+            Self::ProximityRanked => "ts_rank_cd(",
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct FulltextConfig {
+    pub language: FulltextLanguage,
+    pub algorithm: FulltextAlgorithm,
+}
+
+pub struct FulltextDefinition {
+    pub config: FulltextConfig,
+    pub included_fields: HashSet<String>,
+    pub name: String,
+}
+
+impl From<&Directive> for FulltextDefinition {
+    // Assumes the input is a Fulltext Directive that has already been validated because it makes
+    // liberal use of unwrap() where specific types are expected
+    fn from(directive: &Directive) -> Self {
+        let name = directive
+            .argument("name")
+            .unwrap()
+            .as_string()
+            .unwrap()
+            .clone();
+
+        let algorithm = FulltextAlgorithm::try_from(
+            directive.argument("algorithm").unwrap().as_enum().unwrap(),
+        )
+        .unwrap();
+
+        let language =
+            FulltextLanguage::try_from(directive.argument("language").unwrap().as_enum().unwrap())
+                .unwrap();
+
+        let included_entity_list = directive.argument("include").unwrap().as_list().unwrap();
+        // Currently fulltext query fields are limited to 1 entity, so we just take the first (and only) included Entity
```

Comment:
Yes, I believe that aside from updates to validation it'd be mostly contained to query side changes.  I'd like to reread our interface implementation and align with @lutter before planning since they will be similar, but to give you an idea of where we'll likely go: 
- We'd still use the same table structure in which each entity table includes an additional field for each fulltext query field.  These fulltext fields are updated as rows in the specific entity table are updated.  For a fulltext field that covers multiple entities, the field well have columns in each of the respective entity tables and at query time the tables will be joined together and the text search vectors (fulltext field data) concatenated together before applying the text search and filtering. We'll need to update the fulltext data that is stored in the Layout to support query generation that joins all relevant entity tables together. 
- The result type for fulltext query fields will be an interface which will be added to the schema on deployment. It will also likely be represented as an entity on the schema, so query generation can proceed similarly to other entity queries.

Reply:
Yes, that all matches my expectations. 👍